http/src/v0/pin.rs

```diff
@@ -116,26 +107,224 @@ impl<'a> TryFrom<&'a str> for AddRequest {
 
-        Ok(AddRequest {
-            args,
-            recursive: recursive.unwrap_or(false),
-            progress: progress.unwrap_or(false),
+        Ok(ListRequest {
+            arg: args,
+            filter: filter.unwrap_or(PinFilter::All),
+            quiet: quiet.unwrap_or(false),
+            // this default was mentioned in the pin/ls api
+            stream: quiet.unwrap_or(true),
+            timeout,
         })
     }
 }
 
-/// Filter to perform custom `warp::query::<AddRequest>`. This needs to be copypasted around as
-/// HRTB is not quite usable yet.
-fn add_request() -> impl Filter<Extract = (AddRequest,), Error = Rejection> + Clone {
+#[derive(Debug, serde::Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum PinFilter {
+    Direct,
+    Indirect,
+    Recursive,
+    All,
+}
+
+impl Default for PinFilter {
+    fn default() -> Self {
+        PinFilter::All
+    }
+}
+
+impl PinFilter {
+    fn to_mode(&self) -> Option<PinMode> {
+        use PinFilter::*;
+        match self {
+            Direct => Some(PinMode::Direct),
+            Indirect => Some(PinMode::Indirect),
+            Recursive => Some(PinMode::Recursive),
+            All => None,
+        }
+    }
+}
+
+/// `pin/ls` per https://docs.ipfs.io/reference/http/api/#api-v0-pin-ls
+pub fn list<T: IpfsTypes>(
+    ipfs: &Ipfs<T>,
+) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
+    with_ipfs(ipfs).and(list_options()).and_then(list_inner)
+}
+
+fn list_options() -> impl Filter<Extract = (ListRequest,), Error = Rejection> + Clone {
     warp::filters::query::raw().and_then(|q: String| {
-        let res = AddRequest::try_from(q.as_str())
+        let res = ListRequest::try_from(q.as_str())
             .map_err(StringError::from)
             .map_err(warp::reject::custom);
 
         futures::future::ready(res)
     })
 }
+
+async fn list_inner<T: IpfsTypes>(
+    ipfs: Ipfs<T>,
+    req: ListRequest,
+) -> Result<impl Reply, Rejection> {
+    use futures::stream::{StreamExt, TryStreamExt};
+
+    #[derive(serde::Serialize)]
+    struct Good {
+        #[serde(rename = "Cid")]
+        cid: StringSerialized<Cid>,
+        #[serde(rename = "Type")]
+        mode: Cow<'static, str>,
+    }
+
+    impl From<(Cid, Cow<'static, str>)> for Good {
+        fn from((cid, mode): (Cid, Cow<'static, str>)) -> Good {
+            Good {
+                cid: StringSerialized(cid),
+                mode,
+            }
+        }
+    }
+
+    if req.arg.is_empty() {
+        let st = ipfs.list_pins(req.filter.to_mode()).await;
+
+        if req.stream {
+            let st = st.map_ok(|(cid, mode)| {
+                Good::from((
+                    cid,
+                    match mode {
+                        PinMode::Direct => Cow::Borrowed("direct"),
+                        PinMode::Indirect => Cow::Borrowed("indirect"),
+                        PinMode::Recursive => Cow::Borrowed("recursive"),
```

Comment:
nit: `Cow::Borrowed` is the default outcome of doing `.into()` on static strings (where the `Cow<'a, str>` type is expected)